File: SnapCS/Form2.cs
namespace SnapCS
{
    public partial class Form2 : Form
    {
        public Form2()
        {
            InitializeComponent();

            // カーソルのある方のスクリーンを画像としてキャプチャ
            var bounds = Screen.GetBounds(Cursor.Position);
            var captured = new Bitmap(bounds.Width, bounds.Height);
            using var g = Graphics.FromImage(captured);
            g.CopyFromScreen(bounds.Location, Point.Empty, bounds.Size);

            // 最前面で最大化表示
            this.Text = "";
            this.ControlBox = false;
            this.StartPosition = FormStartPosition.Manual;
            this.Location = bounds.Location;
            this.WindowState = FormWindowState.Maximized;
            this.TopMost = true;
            this.Show();
            this.TopMost = false;

            // なにかキーが押されるととりあえず閉じる
            this.KeyPress += (sender, e) => this.Close();

            // 画像表示、マウスによる矩形選択
            var isSelecting = false;
            var p1 = Point.Empty;
            var selRect = Rectangle.Empty;
            var picturebox1 = new PictureBox {
                Parent = this,
                Size = captured.Size,
                Image = captured,
                Cursor = Cursors.Cross
            };

            picturebox1.MouseDown += (sender, e) => {
                // 選択はじめ
                if (e.Button == MouseButtons.Left) {
                    isSelecting = true;
                    p1 = e.Location;
                }
            };

            picturebox1.MouseMove += (sender, e) => {
                // 選択中
                if (isSelecting) {
                    var p2 = e.Location;
                    selRect = new Rectangle(Math.Min(p1.X, p2.X), Math.Min(p1.Y, p2.Y),
                                            Math.Abs(p1.X - p2.X), Math.Abs(p1.Y - p2.Y));
                    picturebox1.Invalidate();
                }
            };

            picturebox1.MouseUp += (sender, e) => {
                // 選択おわり
                if (e.Button == MouseButtons.Left) {
                    isSelecting = false;

                    if (selRect.Width > 0 && selRect.Height > 0) {
                        // 画像を切り抜いてクリップボードへコピー
                        var cropped = new Bitmap(selRect.Width, selRect.Height);
                        using var g = Graphics.FromImage(cropped);
                        g.DrawImage(picturebox1.Image, new Rectangle(0, 0, cropped.Width, cropped.Height), selRect, GraphicsUnit.Pixel);
                        Clipboard.SetImage(cropped);
                    }
                }
            };

            picturebox1.Paint += (sender, e) => {
                // 選択中の矩形を描画
                using var pen1 = new Pen(Color.Red, 1);
                e.Graphics.DrawRectangle(pen1, selRect);
            };

        }
    }
}

File: SnapVB/Form2.vb
Public Class Form2
    Public Sub New()
        InitializeComponent()

        ' カーソルのある方のスクリーンを画像としてキャプチャ
        Dim bounds = Screen.GetBounds(Cursor.Position)
        Dim captured As New Bitmap(bounds.Width, bounds.Height)
        Using g = Graphics.FromImage(captured)
            g.CopyFromScreen(bounds.Location, Point.Empty, bounds.Size)
        End Using

        ' 最前面で最大化表示
        Me.Text = ""
        Me.ControlBox = False
        Me.StartPosition = FormStartPosition.Manual
        Me.Location = bounds.Location
        Me.WindowState = FormWindowState.Maximized
        Me.TopMost = True
        Me.Show()
        Me.TopMost = False

        ' なにかキーが押されるととりあえず閉じる
        AddHandler Me.KeyPress, Sub(sender, e) Me.Close()

        ' 画像表示、マウスによる矩形選択
        Dim isSelecting = False
        Dim p1 As Point
        Dim selRect As Rectangle
        Dim picturebox1 As New PictureBox With {
            .Parent = Me,
            .Size = captured.Size,
            .Image = captured,
            .Cursor = Cursors.Cross
        }

        AddHandler picturebox1.MouseDown,
            Sub(sender, e)
                ' 選択はじめ
                If e.Button = MouseButtons.Left Then
                    isSelecting = True
                    p1 = e.Location
                End If
            End Sub

        AddHandler picturebox1.MouseMove,
            Sub(sender, e)
                ' 選択中
                If isSelecting Then
                    selRect = New Rectangle(Math.Min(p1.X, e.X), Math.Min(p1.Y, e.Y),
                                            Math.Abs(p1.X - e.X), Math.Abs(p1.Y - e.Y))
                    picturebox1.Invalidate()
                End If
            End Sub

        AddHandler picturebox1.MouseUp,
            Sub(sender, e)
                ' 選択おわり
                If e.Button = MouseButtons.Left Then
                    isSelecting = False

                    If selRect.Width > 0 AndAlso selRect.Height > 0 Then
                        ' 画像を切り抜いてクリップボードへコピー
                        Dim cropped As New Bitmap(selRect.Width, selRect.Height)
                        Using g = Graphics.FromImage(cropped)
                            g.DrawImage(picturebox1.Image, New Rectangle(0, 0, cropped.Width, cropped.Height), selRect, GraphicsUnit.Pixel)
                        End Using
                        Clipboard.SetImage(cropped)
                    End If
                End If
            End Sub

        AddHandler picturebox1.Paint,
            Sub(sender, e)
                ' 選択中の矩形を描画
                Using pen1 As New Pen(Color.Red, 1)
                    e.Graphics.DrawRectangle(pen1, selRect)
                End Using
            End Sub

    End Sub
End Class